Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java

```diff
@@ -0,0 +1,488 @@
+ */
+
+package org.tribuo.math.neighbour.kdtree;
+
+import com.oracle.labs.mlrg.olcut.util.Pair;
+import org.tribuo.math.distance.DistanceType;
+import org.tribuo.math.la.SGDVector;
+import org.tribuo.math.neighbour.NeighboursQuery;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * A k-d tree nearest neighbour query implementation.
+ * <p>
+ * For the original algorithm see:
+ * <pre>
+ * J.L. Bentley "Multidimensional Binary Search Trees Used for Associative Searching", Commun. ACM, Vol 18, Sept. 1975, 509–517
+ * <a href="https://doi.org/10.1145/361002.361007">Multidimensional Binary Search Trees Used for Associative Searching</a>
+ * </pre>
+ * <p>
+ * Some aspects of this implementation, particularly the logic which finds the median and partitions the points for each
+ * dimension, are derived from:
+ * <pre>
+ * G. Heineman, G. Pollice, and S. Selkow. "Algorithms in a Nutshell" O'Reilly Media, Inc. 2008.
+ * <a href="http://oreilly.com/catalog/9780596516246/">Algorithms in a Nutshell</a>
+ * </pre>
+ */
+public class KDTree implements NeighboursQuery {
+
+    private final SGDVector[] data;
+    private final int numThreads;
+
+    private final DimensionNode root;
+
+    /**
+     * Constructs a k-d tree nearest neighbour query object using the supplied parameters.
+     *
+     * @param data The data that will be used for neighbour queries.
+     * @param distanceType The distance function.
+     * @param numThreads The number of threads to be used to parallelize queries for multiple points.
+     */
+    KDTree(SGDVector[] data, DistanceType distanceType, int numThreads) {
+        this.data = data;
+        this.numThreads = numThreads;
+
+        int numDimensions = data[0].size();
+
+        // The original data is copied into this array of records. The new array will be rearranged, throughout the
+        // process of tree generation, and the original indices need to maintained.
+        IntAndVector[] points = new IntAndVector[data.length];
+        Stream<Integer> intStream = IntStream.range(0, data.length).boxed();
+        intStream.forEach(i -> points[i] = new IntAndVector(i, data[i]));
+
+        // The dimension comparators that compare points in their specific dimension
+        DimensionComparator[] comparators = new DimensionComparator[numDimensions];
+        for (int i=0; i < numDimensions; i++) {
+            comparators[i] = new DimensionComparator(i);
+        }
+
+        // Start generating the tree at dimension 1, which is index 0 of the array of IntAndVectors.
+        root = generateTree(1, numDimensions, points, 0, data.length - 1, comparators, distanceType);
```

Comment:
When developing this feature, conceptually I was thinking in terms of dimensions instead of features. Dimension 1 - points on a line, Dimension 2 - x,y points in a plane, Dimension 3 - x,y,z points in space, etc. If this seems confusing, the logic could be adjusted to start at (feature) 0, instead of (dimension) 1.

Reply:
Ok, that makes sense, but given the indices are numbering the dimensions I think it would be better to start that numbering at zero the same way we do with features.

Reply:
Sure, that is an easy change. I actually like it better this way too.